### PR TITLE
Support for Pusher Beams (iOS)

### PR DESCRIPTION
**Pull request by emilkarl**
Added support for the new Pusher Beams notifications via [pusher/push-notifications-swift](https://github.com/pusher/push-notifications-swift).

Added new information in readme on what to do to include Carthage frameworks. 

Discussed in #21 with @lukabratos and @b8ne

### CODE REVIEW

**Comment by darringrove** [comment]
I would suggest renaming setAppKey (old Pusher term) to setInstanceId (new Beams term).

File: ios/RNPusherPushNotifications.m [hunk 1]
```diff
@@ -21,45 +22,46 @@ - (dispatch_queue_t)methodQueue
...
     });
 }
 
 RCT_EXPORT_METHOD(subscribe:(NSString *)interest)
 {
     dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
         RCTLogInfo(@"Subscribing to: %@", interest);
-        [[[self pusher] nativePusher] subscribe:(NSString *)interest];
+        NSError *anyError;
+        [[PushNotifications shared] subscribeWithInterest:interest error:&anyError completion:^{
```
Comment (lukabratos) [inline ios/RNPusherPushNotifications.m]: It would be good if errors would be handled.
We could do something as simple as [this](https://github.com/pusher/push-notifications-swift/blob/master/push-notifications-objc/push-notifications-objc/AppDelegate.m#L21).

**Comment by b8ne** [comment]
@emilkarl legend, thanks heaps for this.  Before I merge though it would be good to look at @lukabratos suggestion on errors.  I'd also like to address the Android integration to beams too.  Sorry I have been flat out so haven't got to looking at it yet, this PR will save me a lot of time so hopefully will get to it in the next week.  PR's are obviously always welcome.

**Comment by emilkarl** [comment]
I've done some updates to the library now and the README. Please take another look. I will not be able to dig into Android at the moment but that should be done of course. @b8ne and @lukabratos

**Comment by emilkarl** [comment]
Ready for review @b8ne and @lukabratos 🎉

**Comment by alfonsosn** [comment]
Wow @emilkarl great job at taking the lead on this. I can test Beams on iOS once you merge this request.

File: README.md [hunk 2]
```diff
@@ -1,14 +1,25 @@
 # react-native-pusher-push-notifications
+
 Manage pusher interest subscriptions from within React Native JS
 
-IMPORTANT!!! This module is intended to complement the default [Pusher setup](https://pusher.com/docs/push_notifications).  This module simply allows that implementation to be accessed directly from React Native JS.
+**IMPORTANT!!!**
+
+This module is intended to complement the default [Pusher Carthage setup](https://docs.pusher.com/beams/ios/sdk-integration#install-from-carthage). This module simply allows that implementation to be accessed directly from React Native JS.
```
Comment (lukabratos) [inline README.md]: I would rephrase this like:

This module is intended to complement the setup with [Carthage](https://pusher.com/docs/push_notifications).

Comment (lukabratos) [inline README.md]: I would remove word `simply` as it might be simple for somebody and also not-so-simple for others.
https://jameshfisher.com/2017/02/22/dont-use-simply

File: README.md [hunk 3]
```diff
@@ -17,39 +28,44 @@ IMPORTANT!!! This module is intended to complement the default [Pusher setup](ht
 
 #### iOS
 
-1. In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
-2. Go to `node_modules` ➜ `react-native-pusher-push-notifications` and add `RNPusherPushNotifications.xcodeproj`
-3. In XCode, in the project navigator, select your project. Add `libRNPusherPushNotifications.a` to your project's `Build Phases` ➜ `Link Binary With Libraries`
-4. Run your project (`Cmd+R`)<
+1.  In XCode, in the project navigator, right click `Libraries` ➜ `Add Files to [your project's name]`
```
Comment (lukabratos) [inline README.md]: XCode -> Xcode

**Comment by emilkarl** [comment]
Xcode warns about not doing `pushNotifications.start`, should we remove this from the js and let this be implemented in the AppDelegate instead? 

```
[RNPusherPushNotifications.m:24] Creating pusher with Instance ID: XXXX
Warning: You should call `pushNotifications.start` from the `AppDelegate.didFinishLaunchingWith`
```

Sometimes I'm getting this when reloading the app. "Shake -> Reload" in react native. Have you ever seen it? Did I miss something about this implementation?

image

Would love to get some feedback on this if you got any? @ewal @lukabratos @b8ne @darringrove @alfonsosn

**Comment by lukabratos** [comment]
Hey

Yes, `start(instanceId:)` should be placed in:
`func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool`

Here's an [example](https://github.com/pusher/push-notifications-swift/blob/master/push-notifications-ios/push-notifications-ios/AppDelegate.swift#L10).

**Comment by alfonsosn** [comment]
@emilkarl I'm trying to run the forked version of your feature branch in my project and I've followed all the instructions that you included, but somehow I keep obtaining the following error:

image

Any idea of what this could be? @lukabratos @emilkarl

**Comment by alfonsosn** [comment]
Hey @emilkarl great job on this. I'm trying to install your branch into my repository for testing, but I'm getting some seems like I'm getting some linking error.

`Showing All Messages
clang: error: no such file or directory: '//Xcode/DerivedData/Project-guqtqduiitskcbafwvozyxsrglvz/Build/Products/Debug-iphoneos/libRNPusherPushNotifications.a'`

Even though I followed the installation instructions and manually linked the xcode project:

image

**Comment by alfonsosn** [comment]
@b8ne and @lukabratos any progress on this PR?

**Comment by emilkarl** [comment]
I'm not sure @alfonsosn. I guess you ran `react-native link`? My `libRNPusherPushNotifications.a` is not red in the project tree. 

Did you look at these things?
https://docs.pusher.com/beams/ios/sdk-integration#install-from-carthage
https://github.com/pusher/push-notifications-swift/issues/65

**Comment by emilkarl** [comment]
Im on vacation and will be back in a couple of weeks. Might be able to help out more then.

**Comment by b8ne** [comment]
Hey @emilkarl just submitted a PR to this branch - wanted to keep the work and conversation in the one place.
I refactored a little but seems to be working for iOS nicely now.
The old version had to pass view instances around so that we could keep track of the main Pusher instance.  With this new version we dont need that though, all we need in the main app delegate now  is the deviceToken and onNotification callbacks.
I then found that when calling RNPusherPushNotifications from the main app delegate it used a different instance to that initialized by RN, so the `bridge` was null and gave an error.  My Objective-C is rubbish so im not sure why, but I found a workaround by seperating the event emitters into a seperate class.
Could you test this new version and if its all good ill merge it.

@lukabratos im almost at the end of my project so will tidy up the Android side in a week or 2 for full support.

**Comment by shazvi** [comment]
hi @b8ne any update on when this will be supported on Android? And will we be tracking it on this thread or a separate one?

Thanks
